Public/Src/Engine/Scheduler/PipExecutor.cs

```diff
@@ -1549,12 +1553,30 @@ private static void MakeSharedOpaqueOutputIfNeeded(IPipExecutionEnvironment envi
                                 }
                             }
 
+                            if (userRetry)
+                            {
+                                counters.IncrementCounter(PipExecutorCounter.ProcessUserRetriesImpactedPips);
+                                if (result.Status == SandboxedProcessPipExecutionStatus.Succeeded)
+                                {
+                                    //TODO update PipsSucceedingAfterUserRetry
+                                }
+                                else if (result.Status == SandboxedProcessPipExecutionStatus.ExecutionFailed)
+                                {
+                                    //TODO update PipsFailingAfterLastUserRetry
+                                }
+                            }
+
                             return result;
                         }
                     });
 
             processExecutionResult.ReportSandboxedExecutionResult(executionResult);

```

Comment:
>ReportSandboxedExecutionResult [](start = 35, length = 30)

`processExecutionResult` is the one that the `Scheduler` gets after executing the process. The method `processExecutionResult.ReportSandboxedExecutionResult` basically include the sandboxed result (i.e., `executionResult`) into `processExecutionResult`. So we don't need to add any method to `IPipExecutionEnvironment`.

Reply:
In the Scheduler.cs, try find the following excerpt:
```
case PipExecutionStep.ExecuteProcess:
                    {
                        MarkPipStartExecuting();

                        if (processRunnable.Weight > 1)
                        {
                            // Only log for pips with non-standard process weights
                            Logger.Log.ProcessPipProcessWeight(loggingContext, processRunnable.Description, processRunnable.Weight);
                        }

                        processRunnable.Executed = true;

                        var executionResult = await worker.ExecuteProcessAsync(processRunnable);

                        // Don't count service pips in process pip counters
                        if (!processRunnable.Process.IsStartOrShutdownKind && executionResult.PerformanceInformation != null)
                        {
                            var perfInfo = executionResult.PerformanceInformation;

                            if (perfInfo != null)
                            {
                                m_groupedPipCounters.AddToCounters(
                                    processRunnable.Process,
                                    new[]
                                    {
                                        (PipCountersByGroup.IOReadBytes, (long)perfInfo.IO.ReadCounters.TransferCount),
                                        (PipCountersByGroup.IOWriteBytes, (long)perfInfo.IO.WriteCounters.TransferCount)
                                    },
                                    new[] { (PipCountersByGroup.ExecuteProcessDuration, perfInfo.ProcessExecutionTime) }
                                );
                            }
                        }
```

In that excerpt we update `m_groupedPipCounters`. We can follow this pattern by adding code there to update the pip properties that you're tracking. I think this is how we propagate values from `PipExecutor` to `Scheduler`.

---
In reply to: [332259811](https://github.com/microsoft/BuildXL/pull/1000#discussion_r332259811) [](ancestors = 332259811)

Reply:
I would then still need a way to pass up if there had been retries in that case.  The main concern is what type of structure should be used to globally collect the information.  The existing counters are either long or int, and have helpers for aggregating correctly across all the machines.  I basically need to do that for PipIds.

Reply:
ExecutionResult can be created from the SandboxedProcessPipExecutionResult, though access to that is restricted (part of the private InnerUnsealedState) and we'll still need to pass through any explicit data we want available in the scheduler.